src/utils/TraceUtils.test.js

```diff
@@ -526,5 +515,152 @@ describe('treeDecorate', () => {
         expect(decorated).not.toHaveProperty('tra.foo', 'bar');
         expect(decorated).not.toHaveProperty('tra.tra.foo', 'bar');
     });
+});
+
+import { forwardMatching, backwardMatching, highlightProperty } from './TraceUtils.js';
+
+describe('forwardMatching, backwardMatching', () => {
+
+    const smallTrace = { cons: 'Cons', num: 1, trace: null };
+    // Small trace is a prefix of long trace.
+    const longTrace = {
+                        cons: 'Cons',
+                        num: 41,
+                        trace: {
+                            cons: 'Cons',
+                            num: 1,
+                            trace: {
+                                cons: 'Cons',
+                                num: 39,
+                                trace: {
+                                    cons: 'Cons',
+                                    num: 1,
+                                    trace: null
+                                }
+                            }
+                        }
+    };
+    const unionTrace = { 'cons': 'Union', trace1: smallTrace, trace2: longTrace };
+    const divergentUnion1 = { cons: 'Union', trace1: null, trace2: smallTrace };
+    const divergentUnion2 = { cons: 'Union', trace1: smallTrace, trace2: null };
+    const divergentUnion3 = {
+        ...smallTrace,
+        trace: { 'cons': 'Union', trace1: smallTrace, trace2: smallTrace }
+    };
+
+    // Abstract the tests to reduce copy-paste.
+    const expectation = (matched, bool1, bool2, bool3) => {
+        expect(matched).toHaveProperty(highlightProperty, false);
+        expect(matched.exps[0]).toHaveProperty(highlightProperty, false);
+        expect(matched.exps[0]).toHaveProperty('exp.' + highlightProperty, bool1);
+        expect(matched.exps[1]).toHaveProperty(highlightProperty, bool2);
+        expect(matched.exps[2]).toHaveProperty(highlightProperty, bool3);
+    }
+
+    // Forward matching.
+
+    const expForward = {
```

Comment:
This object will need to be updated when structuredLang is merged, as the JSON from the compiler will have an entirely new structure.